hyperlink compiler error line numbers

mcast opened the pull request:
I should say I haven't brought up an entire playpen, I only tested it with file:/// and hacking one of the UI buttons to inject canned rustc output.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @huonw (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented on static/web.html:
```diff
@@ -6,6 +6,15 @@
+            function editLine(r1) { var e = editSel(); e.clearSelection(); e.setSelectionAnchor(r1-1,0); e.selectLine() }
+            function editShow(r1,c1, r2,c2) { var e = editSel(); old_range = e.getRange(); e.clearSelection(); e.setSelectionAnchor(r1-1,c1-1); e.selectTo(r2-1,c2-1) }
+            function editGo(r1,c1) { var e = editSel(); e.moveCursorTo(r1-1,c1-1,false); old_range = undefined }
+        </script>
```
Could this be added to the js instead of here in the html?

mcast replied:
Sure.  This was hastiness plus my misunderstanding of where functions can be declared, to be reachable from href="javascript:..."

alexcrichton commented on static/web.js:
```diff
@@ -535,6 +535,10 @@
         }).replace(/run `rustc --explain (E\d\d\d\d)` to see a detailed explanation/g, function(text, code) {
             return "see the <a href=https://doc.rust-lang.org/error-index.html#" + code + ">detailed explanation for " + code + "</a>";
+        }).replace(/^&lt;anon&gt;:(\d+):(\d+):\s+(\d+):(\d+)/mg, function(text, r1,c1, r2,c2) {
+            return "<a onmouseout=\"javascript:editRestore()\" onmouseover=\"javascript:editShow("+r1+","+c1+", "+r2+","+c2+")\" href=\"javascript:editGo("+r1+","+c1+")\">" + text + "</a>";
```
Can this be wrapped a bit to not be such a long line as well?

mcast commented:
Tidier now.  Also discovered some of the new features were broken, fixed now.

I added style and detect panics, but suspect I'm not finding all the possible line number sites. I will "watch" for issues later.

mcast commented:
The "editor.$blockScrolling = Infinity" fix for console noise broke scrolling.  Fixed it better now.

alexcrichton commented on static/web.js:
```diff
@@ -553,6 +575,8 @@
         themes = document.getElementById("themes");
         editor = ace.edit("editor");
         set_result.editor = editor;
+        editor.$blockScrolling = Infinity;
```
What's up with this variable? It looks like it may be a little fishy...

respeccing replied:
That var is also used [here](https://github.com/ajaxorg/ace/blob/759c65090f7ef50743d9ae4f5bfba3703fe2b663/lib/ace/autocomplete/popup.js#L215)

It looks like it's needed to be set in order to avoid showing a warning when calling setValue()  https://github.com/ajaxorg/ace/issues/2499

respeccing replied:
[Current](https://github.com/rust-lang/rust-playpen/commit/88045c4b28140d2dd6afeb34c5ef6174627a46ff) rust-playpen actually gives the warning twice:
![blockcursor](https://cloud.githubusercontent.com/assets/17984080/14235237/7cef2f94-f9f8-11e5-9de8-8742a897f200.png)

alexcrichton commented on static/web.js:
```diff
@@ -530,12 +530,34 @@
+            .replace(/\[(E\d\d\d\d)\]/g,
+                     function(text, code) {
+                         return "[<a href=https://doc.rust-lang.org/error-index.html#" + code + ">" + code + "</a>]";
+                     })
```
Could this formatting stay the same way that it was before?

respeccing replied:
Done https://github.com/respeccing/rust-playpen/commit/60842c391caf323d016996c628e3366bc50a8c7c
(I think)

alexcrichton commented on static/web.js:
```diff
@@ -530,12 +530,34 @@
+            " onmouseover=\"javascript:editShowRegion("+r1+","+c1+", "+r2+","+c2+")\"" +
+            " onmouseout=\"javascript:editRestore()\"" +
+            " class=\"linejump\">" + text + "</a>";
+    }
```
Can you add some small docs to these functions to explain what they're doing?

respeccing replied:
I tried: https://github.com/respeccing/rust-playpen/commit/9eb7803f0a9e29c9ad8382c93a79d55d6b373e9a

alexcrichton commented:
Sorry for the delay! Could you also post a screenshot or two of what this looks like?

panicbit commented:
@alexcrichton [Rawgit demo](https://cdn.rawgit.com/mcast/rust-playpen/ad45b1113d988335e3235f0a21cc6e62e828bdfc/static/web.html)

respeccing commented:
According to that rawgit demo, looks like it breaks(or broke at the time) the colors codes:

Otherwise, it looks good:
I positioned cursor at line 33, to show how hovering works next
![screenshot_2016-04-03_07-12-38](https://cloud.githubusercontent.com/assets/17984080/14230867/b667ae14-f96b-11e5-9a87-584f9ff244ed.png)

The below screenshot is when I hover mouse - temporarily scrolls to line(when I move away, it scrolls back to where cursor is: line 33 like in the above screenshot):
![screenshot_2016-04-03_07-12-48](https://cloud.githubusercontent.com/assets/17984080/14230871/e64c28da-f96b-11e5-89ca-afd4a7a4a682.png)
![screenshot_2016-04-03_07-12-55](https://cloud.githubusercontent.com/assets/17984080/14230873/eb91f2f2-f96b-11e5-8eae-08f42e5509a8.png)


And the following is what happens when I click(permanently move cursor):
![screenshot_2016-04-03_07-13-05](https://cloud.githubusercontent.com/assets/17984080/14230877/fe313ff8-f96b-11e5-90c8-92a736774c6b.png)
![screenshot_2016-04-03_07-13-11](https://cloud.githubusercontent.com/assets/17984080/14230879/01bc63be-f96c-11e5-84e1-e2a4027069b5.png)

respeccing commented:
Nevermind! The colors actually work when this PR is applied on latest master!

![untitled](https://cloud.githubusercontent.com/assets/17984080/14231149/287c029a-f979-11e5-905f-7695217813d1.png)


Should I attempt to address the last of Alex's comments and resubmit this as new PR? (uses original commits - so author preserved) - on it anyway, if there are no objections xD

mcast commented:
Hi @respeccing,
Thanks for picking this up in #192, I guess my side-project-budget dried up but it'll be good to have it merged in some form.
